File: app/src/main/java/com/innov8/memeit/Adapters/UserListAdapter.java
package com.innov8.memeit.Adapters;

import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.innov8.memegenerator.Adapters.ListAdapter;
import com.innov8.memegenerator.Adapters.MyViewHolder;
import com.innov8.memegenerator.loading_button_lib.customViews.CircularProgressButton;
import com.innov8.memegenerator.loading_button_lib.interfaces.OnAnimationEndListener;
import com.innov8.memeit.Activities.ProfileActivity;
import com.innov8.memeit.CustomClasses.FontTextView;
import com.innov8.memeit.CustomViews.ProfileDraweeView;
import com.innov8.memeit.KUtilsKt;
import com.innov8.memeit.R;
import com.memeit.backend.MemeItClient;
import com.memeit.backend.dataclasses.MUser;
import com.memeit.backend.dataclasses.User;
import com.memeit.backend.MemeItUsers;
import com.memeit.backend.OnCompleted;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import okhttp3.ResponseBody;

public class UserListAdapter extends ListAdapter<User> {
    private Typeface tf;
    private MUser myUser;
    float size;
    public UserListAdapter(@NotNull Context mContext) {
        super(mContext, R.layout.list_item_follower);
        tf = Typeface.createFromAsset(getMContext().getAssets(), FontTextView.asset);
        myUser= MemeItClient.INSTANCE.getMyUser();
        size=mContext.getResources().getDimension(R.dimen.profile_mini_size);
    }

    @NotNull
    @Override
    public MyViewHolder<User> createViewHolder(@NotNull View view) {
        return new UserListViewHolder(view);
    }
    private boolean isMe(String id){
        return myUser.getId().equals(id);
    }

    public class UserListViewHolder extends MyViewHolder<User> {
        ProfileDraweeView followerImage;
        TextView followerName;
        TextView followerDetail;
        CircularProgressButton followButton;

        public UserListViewHolder(@NonNull View itemView) {
            super(itemView);
            followerImage = itemView.findViewById(R.id.notif_icon);
            followerName = itemView.findViewById(R.id.follower_name);
            followButton = itemView.findViewById(R.id.follower_follow_btn);
            followerDetail = itemView.findViewById(R.id.follower_detail);
            followButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    String t = followButton.getText().toString();
                    followButton.startAnimation();
                    User user = getItemAt(getItem_position());
                    if (t.equalsIgnoreCase("unfollow")) {
                        MemeItUsers.INSTANCE.unfollowUser(user.getUid()).enqueue(new OnCompleted<ResponseBody>() {
                            @Override
                            public void onSuccess(ResponseBody responseBody) {
                                followButton.revertAnimation(new OnAnimationEndListener() {
                                    @Override
                                    public void onAnimationEnd() {
                                        followButton.setText("Follow");
                                    }
                                });
                            }

                            @Override
                            public void onError(String error) {
                                followButton.revertAnimation();
                                Toast.makeText(getMContext(), "Error " + error, Toast.LENGTH_SHORT).show();
                            }
                        });
                    } else {
                        MemeItUsers.INSTANCE.followUser(user.getUid()).enqueue(new OnCompleted<ResponseBody>() {
                            @Override
                            public void onSuccess(ResponseBody responseBody) {
                                followButton.revertAnimation(new OnAnimationEndListener() {
                                    @Override
                                    public void onAnimationEnd() {
                                        followButton.setText("Unfollow");
                                    }
                                });
                            }

                            @Override
                            public void onError(String error) {
                                followButton.revertAnimation();
                                Toast.makeText(getMContext(), "Error " + error, Toast.LENGTH_SHORT).show();
                            }
                        });
                    }
                }
            });
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent i = new Intent(getMContext(), ProfileActivity.class);
                    i.putExtra("user", getItemAt(getItem_position()));
                    getMContext().startActivity(i);
                }
            });
            followButton.setTypeface(tf);
        }

        public void bind(User user) {
            followerName.setText(user.getName());
            followerDetail.setText(user.getPostCount() + " posts");
            followerImage.setText(KUtilsKt.prefix(user.getName()));
            KUtilsKt.loadImage(followerImage,user.getImageUrl(),size,size);
            if(isMe(user.getUid())){
                followButton.setVisibility(View.GONE);
            }else{
                followButton.setVisibility(View.VISIBLE);
                if (user.isFollowedByMe()) {
                    followButton.setText("Unfollow");
                } else {
                    followButton.setText("Follow");
                }
            }


        }
    }

}

File: app/src/main/java/com/innov8/memeit/adapters/UserListAdapter.kt
package com.innov8.memeit.adapters

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.view.View
import android.widget.TextView
import android.widget.Toast
import com.github.ybq.android.spinkit.style.CubeGrid
import com.innov8.memeit.commons.MyViewHolder
import com.innov8.memegenerator.loading_button_lib.customViews.CircularProgressButton
import com.innov8.memeit.activities.ProfileActivity
import com.innov8.memeit.R
import com.innov8.memeit.commons.SimpleELEListAdapter
import com.innov8.memeit.commons.toast
import com.innov8.memeit.commons.views.ProfileDraweeView
import com.innov8.memeit.utils.formatNumber
import com.innov8.memeit.utils.loadImage
import com.innov8.memeit.utils.prefix
import com.memeit.backend.MemeItClient
import com.memeit.backend.MemeItClient.context
import com.memeit.backend.MemeItUsers
import com.memeit.backend.OnCompleted
import com.memeit.backend.call
import com.memeit.backend.models.Meme
import com.memeit.backend.models.MyUser
import com.memeit.backend.models.User
import okhttp3.ResponseBody

class UserListAdapter(mContext: Context, override var emptyDescription: String) : SimpleELEListAdapter<User>(mContext, R.layout.list_item_follower) {
    override var emptyDrawableId: Int = R.drawable.empty_list
    override var errorDrawableId: Int = R.drawable.ic_no_internet
    override var errorDescription: String = "Couldn't load Memes"
    override var emptyActionText: String? = ""
    override var errorActionText: String? = "Try Again"
    override val loadingDrawable = CubeGrid().apply {
        color = Color.rgb(255, 100, 0)
    }
    private val myUser: MyUser? = MemeItClient.myUser

    init {
        if (onItemClicked == null)
            onItemClicked = {
                val i = Intent(context, ProfileActivity::class.java)
                i.putExtra("user", it)
                context.startActivity(i)
            }
    }

    override fun createViewHolder(view: View): MyViewHolder<User> {
        return UserListViewHolder(view)
    }

    internal fun isMe(id: String?): Boolean {
        return myUser!!.id == id
    }

    private fun getItemByID(id: String) = items.find { it.uid == id }

    inner class UserListViewHolder(itemView: View) : MyViewHolder<User>(itemView) {
        private val followerImage: ProfileDraweeView = itemView.findViewById(R.id.notif_icon)
        private val followerName: TextView = itemView.findViewById(R.id.follower_name)
        private val followerDetail: TextView = itemView.findViewById(R.id.follower_detail)
        private val followV: TextView = itemView.findViewById(R.id.follow_user)

        init {

            followV.visibility = View.VISIBLE
            followV.setOnClickListener { _ ->
                val uid = items[item_position].uid!!
                if (followV.text == "Unfollow") {
                    followV.text = "Unfollowing..."
                    MemeItUsers.unfollowUser(uid).call({
                        followV.text = "Follow"
                        getItemByID(uid)?.isFollowedByMe = false
                    }, {
                        context.toast("Failed to Unfollow")
                        followV.text = "Unfollow"
                    })
                } else {
                    followV.text = "Following..."
                    MemeItUsers.followUser(uid).call({
                        followV.text = "Unfollow"
                        getItemByID(uid)?.isFollowedByMe = true
                    }, {
                        context.toast("Failed to Follow")
                        followV.text = "Follow"
                    })
                }
            }

            itemView.setOnClickListener {
                onItemClicked?.invoke(items[item_position])
            }
        }

        override fun bind(t: User) {
            followerName.text = t.name
            followerDetail.text = t.postCount.formatNumber("post", "posts")
            followerImage.setText(t.name.prefix())
            followerImage.loadImage(t.imageUrl)
            if (isMe(t.uid!!)) {
                followV.visibility = View.GONE
            } else {
                followV.visibility = View.VISIBLE
                followV.text = if (t.isFollowedByMe) "Unfollow" else "Follow"
            }


        }
    }

}
